### PR TITLE
rework build system

### PR DESCRIPTION
**Pull request by paul-marechal**
#### What it does

Fixes: https://github.com/eclipse-theia/theia/issues/7275.

Simplify `scripts/compile-references.js`. TypeScript Server now properly
understand TypeScript references for navigation, so we can simplify our
`tsconfig` setup.

Rewrite some `npm/yarn` scripts to do less on install (`yarn` or `yarn
install`). It will only run the required `prepare` scripts across our
packages, wire `tsconfig.json` files and build the TypeScript sources.

To build the example applications you need to run one of `yarn browser
build`, `yarn electron build` or `yarn build:examples`.

Linting is done as part of CI, but it won't automatically run locally
anymore. You can run `yarn lint` to lint your code. For the best
experience you should also enable the ESLint extension in your IDE.

#### How to test

Installation:

- `yarn` alone should not build the example applications anymore, nor should it lint.

Building:

- `yarn build` should build TypeScript and both apps.
- `yarn compile` should build TypeScript.
- `yarn browser build` should build both TypeScript and the browser app.
- `yarn electron build` should build both TypeScript and the electron app.
- `yarn build:examples` should build both apps.

Watching:

- `yarn watch` should watch TypeScript and both example apps.
- `yarn watch:ts` should watch only TypeScript.
- `yarn browser watch` should watch both TypeScript and the browser app.
- `yarn electron watch` should watch both TypeScript and the electron app.

Linting:

- `yarn lint` should lint TypeScript sources.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by paul-marechal** [comment]
This is especially useful when using the main repo as a submodule for a different repository such as a Theia extension:

My approach was to clone Theia as a submodule in the extension repo, then add `submodules/theia/packages/*` and `submodules/theia/dev-packages/*` to the parent repo yarn workspaces. Finally, I would run the TypeScript compilation in Theia to simply build the TypeScript packages, as this is all that is required for dependents to consume.

`yarn link` doesn't play nicely with our Theia building tools for some reason, hence why this alternative approach is interesting.

This PR also makes developing a bit more comfortable: only the minimal amount of work required is done, if you need more you need to explicitly trigger it. E.g. building a given app, linting, etc...

File: doc/Developing.md
Comment (colin-grant-work) [inline doc/Developing.md:193]: The only thing that gives me (slight) pause here is the change in behavior of the plain `yarn` command. If we want to have more options in how we build, there are basically two options: reduce `yarn`'s functionality and create a new command that has its old functionality (this PR), or leave `yarn` as was and create a new script with the reduced functionality of installing and transpiling. I don't think it matters much either way. Either we tell new people 'just `yarn` and then start' or we tell new people 'just `yarn build` and then start.' I might incline toward leaving `yarn` alone so that casual users' workflows don't change, but there's a certain iconicity to the shorter command (`yarn`) do less work than the longer command (`yarn build`).

Comment (paul-marechal) [reply doc/Developing.md:193]: My biggest pain is that `yarn` alone is a shortcut to `yarn install` meaning it is meant to deal with dependencies. At best it is common practice to prepare the packages by transpiling sources, but linting and building two applications is overkill. This is especially annoying to work with when we need to work with dependencies when you have to keep an eye at the `yarn` output to see when to `ctrl+c` it.

With this PR, doing `yarn` will at least put the project in a state that can be published/consumed (we don't need to build the example apps to publish).

Comment (paul-marechal) [reply doc/Developing.md:193]: More importantly: `yarn (install)` is the first command you HAVE to run in order to work with any npm package. Making that command do everything means there is no way to do less when preferable to do so.

Comment (colin-grant-work) [reply doc/Developing.md:193]: Sounds fair to me. 👍

**Comment by colin-grant-work** [comment]
> This is especially useful when using the main repo as a submodule for a different repository such as a Theia extension

This sounds it goes some way to addressing some of the pain points described in #9153, as well.

**Comment by tsmaeder** [comment]
> yarn link doesn't play nicely with our Theia building tools for some reason, hence why this alternative approach is interesting.

Could you detail in what way it does not play well with `yarn link`? it seems to work for me, but I fear I'm missing something.

File: scripts/compile-references.js
Comment (tsmaeder) [inline scripts/compile-references.js:41]: I don't quite understand why we need to rewrite references as part of the build at all: aren't they static and inferred by the usage relation between packages?

File: package.json
Comment (tsmaeder) [inline package.json:67]: Where did these go?

File: package.json
Comment (tsmaeder) [inline package.json:49]: I kind of regret that we don't have a single command anymore that will "build everyting" after checking out the source. As I understand it, we'll have to `yarn install && yarn build` now, right?

File: doc/Developing.md
Comment (paul-marechal) [reply doc/Developing.md:261]: Then there's nothing to watch?

Do we have such non-TypeScript package?

File: examples/browser/package.json
Comment (paul-marechal) [reply examples/browser/package.json:66]: Good catch: it is a typo and should be `rebuild:browser`, it is to ensure native dependencies are built for Node and not Electron.

File: scripts/compile-references.js
Comment (paul-marechal) [reply scripts/compile-references.js:60]: `main` is usually a function entry point, like C/C++/C#/Java etc... If I could I would use top-level await but our supported Node version doesn't supports this yet.

File: package.json
Comment (paul-marechal) [reply package.json:49]: I can add a `yarn everything` or something similiar, otherwise yes you have to do what you said.

File: package.json
Comment (paul-marechal) [reply package.json:152]: Good catch, I think yarn did that because I sure didn't, I'll remove those.

File: doc/Developing.md
Comment (paul-marechal) [reply doc/Developing.md:261]: This mostly refers to TypeScript compilation which happens before bundling, so if we have a pure JS extension then there will be nothing to watch.

File: scripts/compile-references.js
Comment (paul-marechal) [reply scripts/compile-references.js:41]: They are static and inferred. The reason we do it on build is to make sure everything is in sync since building relies on it.

File: scripts/compile-references.js
Comment (colin-grant-work) [reply scripts/compile-references.js:60]: Maybe just name it the same thing as the module, in camelCase (`compileReferences`), the same way we (often) name modules the same thing as the main class they export?

**Comment by paul-marechal** [comment]
> Could you detail in what way it does not play well with `yarn link`? it seems to work for me, but I fear I'm missing something.

All the times I remember doing `yarn link` in each Theia package for consumption in third-party Theia extension projects, it ended up with Webpack getting confused and duplicating packages in the output bundles for some reason, leading to runtime errors.

Since we migrated to Webpack 5 maybe this won't be an issue anymore...

So independently from that, this PR still simplifies the `compile-references.js` script, allows one to manage dependencies without triggering a full build, and cleans up scripts.

File: packages/callhierarchy/tsconfig.json
Comment (paul-marechal) [reply packages/callhierarchy/tsconfig.json:19]: `callhierarchy` depends on `monaco` as of master already:

https://github.com/eclipse-theia/theia/blob/faae393f66c779c9640c77b0d23c60a4d0d10f62/packages/callhierarchy/package.json#L8
https://github.com/eclipse-theia/theia/blob/faae393f66c779c9640c77b0d23c60a4d0d10f62/packages/callhierarchy/compile.tsconfig.json#L19

File: scripts/compile-references.js
Comment (paul-marechal) [reply scripts/compile-references.js:60]: And yes, making I/O stuff concurrent always yields better performance. Everything was sync before, and as part of simplifying the script I made it async where it made sense.

File: scripts/compile-references.js
Comment (paul-marechal) [reply scripts/compile-references.js:41]: Note that it won't write anything on disk if everything is already synced.

File: package.json
Comment (paul-marechal) [reply package.json:67]: In their respective example applications. I just tweaked the `rebuild` scripts to better handle the use case when it is invoked from non-root locations.

File: examples/browser/package.json
Comment (paul-marechal) [reply examples/browser/package.json:66]: Ah not a typo, there's a `rebuild` script in the same `package.json` now.

File: package.json
Comment (paul-marechal) [reply package.json:49]: Added a `yarn all` command that does everything again.

File: package.json
Comment (paul-marechal) [reply package.json]: Turns out this was due to an aborted `theia rebuild:electron` command: it used to modify the current `package.json` to pass arguments to `electron-rebuild`... Fixed it to not do that anymore, and rather use CLI options.

File: doc/Developing.md
Comment (paul-marechal) [reply doc/Developing.md:261]: I understand `TypeScript package` as a package written using TypeScript and requiring transpilation.

Comment (tsmaeder) [reply doc/Developing.md:258]: So it's really "typescript files", because some packages may container both *.js and *.ts files, right? Also, not sure how that works, exactly, but don't we have to reload the code when *.js files change, as well?

File: doc/Developing.md
Comment (tsmaeder) [inline doc/Developing.md:268]: "Everything" being both the regular packages and the example applications?

File: examples/browser/package.json
Comment (tsmaeder) [reply examples/browser/package.json:68]: But was _not_ doing some kind of "rebuild:browser" a mistake in the old code?

File: package.json
Comment (tsmaeder) [reply package.json:67]: So it would be `yarn browser rebuild` now, right.

File: packages/callhierarchy/tsconfig.json
Comment (tsmaeder) [reply packages/callhierarchy/tsconfig.json:19]: I think the dependency is there by mistake, but I agree it's not your problem in this PR.

File: scripts/compile-references.js
Comment (tsmaeder) [reply scripts/compile-references.js:41]: So I _did_ understand that right. It's a pity the tsc compiler can't infer those if we tell it what the set of source packages are 🤷  Fair enough.

File: scripts/compile-references.js
Comment (tsmaeder) [reply scripts/compile-references.js:60]: > And yes, making I/O stuff concurrent always yields better performance

Well, that's only true if you can do something in the meanwhile that is not I/O. But anyway...async code _is_ more complex to understand at run time, which is a cost to be measured against the gains. `await` makes it _look_ simple, but it isn't.

Comment (tsmaeder) [reply scripts/compile-references.js:60]: I like @colin-grant-work's idea about the "script-named" top functions.

File: doc/Developing.md
Comment (paul-marechal) [reply doc/Developing.md:258]: In Theia the watch scripts are related to code transformations. There's basically two steps: TypeScript compilation and Webpack bundling. This documentation refers to the former. Javascript sources don't require any change during the first step as well, but bundling still depend on them indeed.

File: examples/browser/package.json
Comment (paul-marechal) [reply examples/browser/package.json:68]: Not so much a mistake as something that I though I could improve here :)

File: package.json
Comment (paul-marechal) [reply package.json:67]: Yes.

File: scripts/compile-references.js
Comment (paul-marechal) [reply scripts/compile-references.js]: I prefer `main` for this kind of use case, but I still renamed it to `compileTypeScriptReferences`.

**Comment by msujew** [comment]
@paul-marechal I'll review this, if you're still interested in getting this merged. Would you mind rebasing and fixing merge conflicts?

**Comment by paul-marechal** [comment]
@msujew thanks for proposing! I just rebased. Note that one of the test lifecycle keeps failing and I don't know why...

**Comment by paul-marechal** [comment]
@msujew thanks for the idea! I finally addressed your comments.

I tried improving the GitHub workflows too but I'll keep that for another PR.

File: .gitpod.yml
Comment (msujew) [inline .gitpod.yml:15]: I was wondering why my Theia installation in gitpod didn't run any extensions. I think we should download extensions here as well:

```suggestion
  - init: yarn --network-timeout 100000 && yarn build:examples && yarn download:plugins
```

Comment (paul-marechal) [reply .gitpod.yml]: Ah good catch!

**Review by msujew — APPROVED** [review]
I like the change. I could not find any build-regressions and the new watch options improve my personal development experience 👍

**Comment by paul-marechal** [comment]
There seems to be an issue when rebuilding native dependencies for Electron where `electron-rebuild` doesn't rebuild `@theia/node-pty` for some reason, will fix before merging.

**Comment by paul-marechal** [comment]
We decided to update `electron-rebuild` now while we are improving the rebuild scripts, but this ended up pulling a new dependency requiring a CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23694

**Comment by tsmaeder** [comment]
@paul-marechal That's why I usually push back on suggestions to do "one more useful thing" in my PR's. Small, incremental changes that make the code base better is my ideal. As long as the change in the PR stands up on its own, pushing back on change requests is fine by me. Do followups, I trust you to do them :-)

**Comment by paul-marechal** [comment]
@tsmaeder upgrading `electron-rebuild` is something we'll want to do anyway. But I can extract this is its own PR for sure.

**Comment by tsmaeder** [comment]
@paul-marechal are you aware the tests are failing?

**Comment by paul-marechal** [comment]
@tsmaeder painfully yes, I am yet to find why...

File: CHANGELOG.md
Comment (vince-fugnitto) [inline CHANGELOG.md:88]: Revert?

File: doc/Developing.md
Comment (vince-fugnitto) [inline doc/Developing.md:94]: Given that it does not automatically happen anymore, do we need to mention to download plugins?

**Review by vince-fugnitto — COMMENTED** [review]
I confirmed the behavior works well:
- building
- starting both browser and electron apps
- rebuilding native modules works well on Linux

**Review by vince-fugnitto — APPROVED** [review]
The changes look good to me :+1: 
Don't forget to squash the 17 commits on merge ;)

**Comment by paul-marechal** [comment]
@tsmaeder @msujew last heads up before merging. If you guys are still fine with the changes then let's break everyone's muscle memory... I'll open and pin an issue to mention the new commands to try and help with the new workflows.

**Review by msujew — APPROVED** [review]
@paul-marechal Still fine with merging, everything looks good 👍

**Comment by paul-marechal** [comment]
I'll just wait on CI to turn green...

**Comment by thegecko** [comment]
Really miss the easy `yarn start:browser` command in the root to ensure native modules are built correctly and start the browser example.

**Comment by msujew** [comment]
@thegecko I think you should be able to run `yarn browser start` to do exactly the same as the old `yarn start:browser`.

**Comment by thegecko** [comment]
> I think you should be able to run `yarn browser start`

Confirmed, thanks!

**Pull request merged**
